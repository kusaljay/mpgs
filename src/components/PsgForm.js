import React from 'react';
import {IoIosContact} from 'react-icons/io';

const PsgForm = ({inputChanged, submitGenerate, errors}) => {
  return (
    <div className="card content-container">
      <form onSubmit={submitGenerate}>
        <h2 className="psg-heading-card"><IoIosContact className="icon"/>Employee info</h2>
        {errors && <div className="alert alert-danger form-alerts">{errors}</div>}
        <div className="d-flex flex-wrap mr-n4">
          <div className="form-group empinfo-field">
            <label>First Name</label>
            <input type="text" className="form-control" name="firstName" onChange={inputChanged} />
          </div>
          <div className="form-group empinfo-field">
            <label>Last Name</label>
            <input type="text" className="form-control" name="lastName" onChange={inputChanged} />
          </div>
          <div className="form-group empinfo-field">
            <label>Annual Salary</label>
            <input type="text" className="form-control dollar" name="annualIncome" onChange={inputChanged} />
          </div>
          <div className="form-group empinfo-field">
            <label>Superannuation Rate</label>
            <input type="text" className="form-control" name="superPercent" onChange={inputChanged} />
          </div>
          <button type="submit" className="btn btn-success btn-post-login">Generate Payslip</button>
        </div>

      </form>
    </div>
  )
}

export default PsgForm;
