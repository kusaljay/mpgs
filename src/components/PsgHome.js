import React from 'react'

const PsgHome = () => {
  return (
    <div>
      <h5>psg header</h5>
    </div>
  )
}

export default PsgHome;
