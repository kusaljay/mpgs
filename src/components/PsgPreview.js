import React from 'react';
import {IoIosContact} from 'react-icons/io';

const PsgPreview = ({employeeData, clickPay, clickAnother}) => {
  return (
    <div>
      <div className="preview-wrapper card content-container">
        <h2 className="psg-heading-card"><IoIosContact className="icon"/>Pay slip for {employeeData.firstName} {employeeData.lastName}</h2>
        {employeeData.dbSubmitSuccessMsg && <div className="alert alert-success form-alerts">{employeeData.dbSubmitSuccessMsg}</div>}
        {employeeData.dbSubmitErrorMsg && <div className="alert alert-danger form-alerts">{employeeData.dbSubmitErrorMsg}</div>}
        <div className="d-flex flex-wrap">
          <table className="table table-psg">
            <thead>
              <tr>
                <th scope="col" className="w-30">Item</th>
                <th scope="col" className="w-70">Emp Details</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td className="w-30">Employee</td>
                <td className="w-70 =">{employeeData.firstName}&nbsp;{employeeData.lastName}</td>
              </tr>
              <tr>
                <td className="w-30">Pay Date</td>
                <td className="w-70">{employeeData.payDate}</td>
              </tr>
              <tr>
                <td className="w-30">Pay Frequency</td>
                <td className="w-70">{employeeData.payFrequency}</td>
              </tr>
              <tr>
                <td className="w-30">Annual Income</td>
                <td className="w-70 dollar">{employeeData.annualIncome}</td>
              </tr>
              <tr>
                <td className="w-30">Gross Income</td>
                <td className="w-70 dollar">{employeeData.grossIncome}</td>
              </tr>
              <tr>
                <td className="w-30">Income Tax</td>
                <td className="w-70 dollar">{employeeData.incomeTax}</td>
              </tr>
              <tr>
                <td className="w-30">Net Income</td>
                <td className="w-70 dollar">{employeeData.netIncome}</td>
              </tr>
              <tr>
                <td className="w-30">Super</td>
                <td className="w-70 dollar">{employeeData.superAnnuation}</td>
              </tr>
              <tr>
                <td className="w-30">Pay</td>
                <td className="w-70 dollar">{employeeData.pay}</td>
              </tr>
            </tbody>
          </table>
          {employeeData.dbSubmitSuccess && <button type="button" onClick={clickAnother} className="btn btn-success btn-post-login add-another">Add another employee</button>}
          {employeeData.showPreview && <button type="button" onClick={clickPay} className="btn btn-success btn-post-login pay">Pay</button>}
        </div>
      </div>
    </div>
  )
}

export default PsgPreview;