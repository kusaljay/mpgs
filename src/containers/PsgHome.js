import React, { Component } from 'react';
import firebase from '../utils/firebase';
import PsgHeader from '../templates/PsgHeader';
import PsgFooter from '../templates/PsgFooter';
import PsgForm from '../components/PsgForm';
import PsgPreview from '../components/PsgPreview';
import {rounding, calcIncomeTax, currentDate} from '../utils/helpers'


class PsgHome extends Component {
  constructor() {
    super();
    this.state = {
      firstName: '',
      lastName: '',
      annualIncome: null,
      superPercent: null,
      grossIncome: null,
      incomeTax: null,
      netIncome: null,
      superAnnuation: null,
      pay: null,
      showPreview: false,
      payDate: '',
      payFrequency: '',
      formErrors: '',
      dbSubmitSuccess: false,
      dbSubmitSuccessMsg: '',
      dbSubmitError: false,
      dbSubmitErrorMsg: '',
      buttonTxt: '',
      employees: []
    }
    this.initialState = {...this.state};
  }

  handleInputChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value 
    });
  }
  
  handleValidation = () => {
    const {firstName, lastName, annualIncome, superPercent} = this.state;
    let errors = '';

    if ( firstName === '' || lastName === '' || annualIncome === null || superPercent === null ) {
      errors = 'All fields are mandatory.';
      this.setState({formErrors: errors});
      return false;
    }
    if ( annualIncome < 0 || isNaN(annualIncome) || superPercent < 0 || isNaN(superPercent) ) {
      errors = 'Annual Salary and Superannuation should be positive numbers.';
      this.setState({formErrors: errors});
      return false;
    }
    if ( superPercent > 50 ) {
      errors = 'Superannuation Rate should be between 0% - 50%.';
      this.setState({formErrors: errors});
      return false;
    }
    return true;
  }

  generatePaySlip = () => {
    const annualIncome = this.state.annualIncome,
          grossIncome = rounding(annualIncome / 12),
          incomeTax = calcIncomeTax(annualIncome),
          netIncome = grossIncome - incomeTax,
          superAnnuation = rounding((grossIncome * this.state.superPercent) / 100),
          pay = rounding(netIncome - superAnnuation),
          payDate = currentDate(),
          payFrequency = 'Monthly';

    this.setState({
          annualIncome: annualIncome,
          grossIncome: grossIncome, 
          incomeTax: incomeTax, 
          netIncome: netIncome, 
          superAnnuation: superAnnuation, 
          pay: pay,
          payDate: payDate,
          payFrequency: payFrequency
        });
  }
  
  handleSubmit = (e) => {
    e.preventDefault();
    if (this.handleValidation()) {
      this.setState({showPreview: true, buttonTxt: 'Pay'});
      this.generatePaySlip();
    }
  }

  handlePay = (e) => {
    e.preventDefault();
    const employeesRef = firebase.database().ref('/employees');
    const currentEmployee = {
            firstName: this.state.firstName,
            lastName: this.state.lastName,
            payDate: this.state.payDate,
            payFrequency: this.state.payFrequency,
            annualIncome:this.state.annualIncome,
            superPercent: this.state.superPercent,
            grossIncome: this.state.grossIncome,
            incomeTax: this.state.incomeTax,
            netIncome: this.state.netIncome,
            superAnnuation: this.state.superAnnuation,
            pay: this.state.pay
    };
    employeesRef.push(currentEmployee)
    .then(() => {
      const success = `Pay info succefully added for ${this.state.firstName} ${this.state.lastName}`;
      this.setState({dbSubmitSuccess: true, dbSubmitSuccessMsg: success, buttonTxt: `Add another employee`});
    })
    .catch((error) => {
      this.setState({dbSubmitError: true, dbSubmitErrorMsg: error.message});
    });
  }

  handleAddAnother = (e) => {
    e.preventDefault();
    this.setState(this.initialState);
  }

  componentDidMount() {
    /* Use this to view the database getting updated using console.log
       Didn't attempt to render the list employees being added to the database. 
    */
    const employeesRef = firebase.database().ref('/employees');
    employeesRef.on('value', (snapshot) => {
      let empFromDb = snapshot.val();
      console.log(empFromDb);
    });
  }

  renderContent = () => {
    if(!this.state.showPreview) {
      return <PsgForm inputChanged={this.handleInputChange} submitGenerate={this.handleSubmit} errors={this.state.formErrors} />
    } else {
      return <PsgPreview employeeData={this.state} clickPay={this.handlePay} clickAnother={this.handleAddAnother} />
    }
  }
  
  render() {
    return (
      <div className="post-login-main-wrapper post-login">
        <PsgHeader />
        <main className="container">
          <h1 className="psg-heading-main mt-5 mb-2">Pay Slip Generator</h1>
          {this.renderContent()}
        </main>
        <PsgFooter />
      </div>
    );
  }
}



export default PsgHome;
