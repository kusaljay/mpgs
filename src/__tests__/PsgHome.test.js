import React from 'react';
import { shallow } from 'enzyme';
import PsgHome from '../components/PsgHome';
import PsgForm from '../components/PsgForm';
import PsgPreview from '../components/PsgPreview';

let wrapper;

beforeEach(() => {
  wrapper = shallow(<PsgHome />);
});

describe ('PsgHome', () => {
  it ('Should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it ('Should render form', () => {
    wrapper.setProps({showPreview: false});
    expect(wrapper.find(PsgForm)).toHaveLength(1);
  });

  /* it ('Should render preview', () => {
    wrapper.setProps({showPreview: true});
    expect(wrapper.find(PsgPreview)).toHaveLength(1);
  }); */

});