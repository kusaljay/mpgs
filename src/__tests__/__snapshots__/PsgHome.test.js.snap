// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PsgHome Should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PsgHome />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <PsgHeader />,
        <main
          className="container"
        >
          <h1
            className="psg-heading-main mt-5 mb-2"
          >
            Pay Slip Generator
          </h1>
          <PsgForm
            errors=""
            inputChanged={[Function]}
            submitGenerate={[Function]}
          />
        </main>,
        <PsgFooter />,
      ],
      "className": "post-login-main-wrapper post-login",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="psg-heading-main mt-5 mb-2"
            >
              Pay Slip Generator
            </h1>,
            <PsgForm
              errors=""
              inputChanged={[Function]}
              submitGenerate={[Function]}
            />,
          ],
          "className": "container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Pay Slip Generator",
              "className": "psg-heading-main mt-5 mb-2",
            },
            "ref": null,
            "rendered": "Pay Slip Generator",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "errors": "",
              "inputChanged": [Function],
              "submitGenerate": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "main",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <PsgHeader />,
          <main
            className="container"
          >
            <h1
              className="psg-heading-main mt-5 mb-2"
            >
              Pay Slip Generator
            </h1>
            <PsgForm
              errors=""
              inputChanged={[Function]}
              submitGenerate={[Function]}
            />
          </main>,
          <PsgFooter />,
        ],
        "className": "post-login-main-wrapper post-login",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="psg-heading-main mt-5 mb-2"
              >
                Pay Slip Generator
              </h1>,
              <PsgForm
                errors=""
                inputChanged={[Function]}
                submitGenerate={[Function]}
              />,
            ],
            "className": "container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Pay Slip Generator",
                "className": "psg-heading-main mt-5 mb-2",
              },
              "ref": null,
              "rendered": "Pay Slip Generator",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "errors": "",
                "inputChanged": [Function],
                "submitGenerate": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "main",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
